mterm/mterm.go

```diff
@@ -175,252 +343,431 @@ func (t *Terminal) GetScreenAsAnsi() []byte {
-					p = p[5:]
-				// 16M
-				case p[0] == 48 && len(p) > 4 && p[1] == 2:
-					t.cstate.flags |= FlagBG
-					t.cstate.BG = Color{byte(p[2]), byte(p[3]), byte(p[4])}
-					p = p[5:]
-				}
+			err := t.cstate.set(p...)
+			return (*Terminal).normal, err
+		case 'u':
+			t.cursorLine = t.saveCursor[0]
+			t.cursorCol = t.saveCursor[1]
+		case 's':
+			t.saveCursor = [2]int{t.cursorLine, t.cursorCol}
+		case 'c':
+			// TODO: {lpf} (comment by copilot: Send device attributes)
+		case 'h':
+			switch p[0] {
+			case 1004:
+				// TODO: Turn focus report ON
+			}
+		case 'l':
+			switch p[0] {
+			case 25: // hide cursor if first rune is '?'
+			case 1:
+				// TODO: Turn cursor keys to application mode OFF
 			}
+		case 't':
+			// TODO: {lpf} (comment by copilot: Window manipulation)
+		case 'r':
+			top, bottom := 1, t.MaxRows
+			getParams(p, &top, &bottom)
+			t.scrollRegion[0] = clamp(top-1, 0, t.MaxRows)
+			t.scrollRegion[1] = clamp(bottom, 0, t.MaxRows)
+			// TODO: this needs some love, it's not working as expected
+			//
+			switch {
+			case len(p) == 0:
+				// fill(t.screen, Cell{})
+				t.cursorLine = 0
+				t.cursorCol = 0
+			case len(p) == 1:
+				t.cursorLine = 0
+				t.cursorCol = 0
+			default:
+				// fill(t.screen, Cell{})
+				// t.cursorLine = 0
+				// t.cursorCol = 0
+			}
+
+		case 'S': // Scrollup
+			n := 1
+			getParams(p, &n)
+			start := t.scrollRegion[0] * t.MaxCols
+			end := t.scrollRegion[1] * t.MaxCols
+			region := t.screen[start:end]
+			copy(region, region[n*t.MaxCols:])
+			fill(region[len(region)-n*t.MaxCols:], Cell{})
+		case 'T': // Scrolldown
+			n := 1
+			getParams(p, &n)
+			start := t.scrollRegion[0] * t.MaxCols
+			end := t.scrollRegion[1] * t.MaxCols
+			region := t.screen[start:end]
+			copy(region[n*t.MaxCols:], region)
+			fill(region[:n*t.MaxCols], Cell{})
+		default:
+			return (*Terminal).normal, fmt.Errorf("unknown CSI: %d %[1]c", r)
+
 		}
-		return (*Terminal).normal
+		return (*Terminal).normal, nil
+	}
+}
+
+func (t *Terminal) nextLine() {
+	t.cursorLine++
+	if t.cursorLine < t.scrollRegion[1]-1 {
+		return
 	}
+	if t.cursorLine == t.scrollRegion[1] {
+		start := (t.scrollRegion[0]) * t.MaxCols
+		end := t.scrollRegion[1] * t.MaxCols
+		region := t.screen[start:end]
+		// move buffer up 1 line
+		copy(region, region[t.MaxCols:])
+		fill(region[len(region)-t.MaxCols:], Cell{})
+		t.cursorLine--
+		t.cellUpdate++
+	}
+	// Replicate odd xterm behaviour of when the cursor is outside of the region
+	// it will not print neither scroll
+	if t.cursorLine == t.MaxRows {
+		t.cursorLine--
+	}
+}
+
+// Updates returns a sequence number that is incremented every time the screen
+// cells are updated
+func (t *Terminal) Updates() int {
+	t.mux.Lock()
+	defer t.mux.Unlock()
+
+	return t.cellUpdate
+}
+
+// DBGStateFn returns the state func name
+func (t *Terminal) DBGStateFn() string {
+	fi := runtime.FuncForPC(reflect.ValueOf(t.stateProc).Pointer())
```

Comment:
Isn't it better to create a function that locks, takes the state and unlocks instead of using reflection?